### PR TITLE
[dagster-embedded-elt][dlt] support usage of DagsterDltResource from ops

### PR DESCRIPTION
**Pull request by cmpadden**
## Summary & Motivation

Support the ability of using the _dlt_ integration from an `@op`; see similar implementation for Sling below:

- https://github.com/dagster-io/dagster/pull/20564/files#diff-7fd29d4a7e3c4781fd2b55054095105ac002207d0daaa593e76ff192fecaaab9R8-R40

If running from an op context the dlt source, pipeline, and translator are retrieved from the `run` function parameters, otherwise they are taken from the Asset metadata.

## How I Tested These Changes

- Introduced unit test

### CODE REVIEW

**Comment by cmpadden** [comment]
* **#21191** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21191?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @cmpadden and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py [hunk 1]
```diff
@@ -99,26 +99,45 @@ def extract_resource_metadata(
...
+            dlt_pipeline = check.inst(first_asset_metadata.get(META_KEY_PIPELINE), Pipeline)
+            dagster_dlt_translator = check.inst(
+                first_asset_metadata.get(META_KEY_TRANSLATOR), DagsterDltTranslator
+            )
+
+        else:
+            if dlt_source is None or dlt_pipeline is None:
+                raise ValueError(
+                    "dlt_source, and dlt_pipeline parameters must be provided in an `op` context"
+                )
```
Comment (rexledesma) [inline python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py:137]: Use check.invariant

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py [hunk 2]
```diff
@@ -99,26 +99,45 @@ def extract_resource_metadata(
     def run(
         self,
         context: Union[OpExecutionContext, AssetExecutionContext],
+        dlt_source: Optional[DltSource] = None,
+        dlt_pipeline: Optional[Pipeline] = None,
+        dagster_dlt_translator: Optional[DagsterDltTranslator] = None,
         **kwargs,
-    ) -> Iterator[Union[AssetMaterialization, MaterializeResult]]:
+    ) -> Generator[Union[MaterializeResult, AssetMaterialization], None, None]:
```
Comment (rexledesma) [inline python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py:106]: stick to `Iterator`

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt_tests/dlt_tests/test_op.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import dlt
+import duckdb
+from dagster import OpExecutionContext, job, op
+from dagster_embedded_elt.dlt import DagsterDltResource
+
+from .dlt_test_sources.duckdb_with_transformer import pipeline
+
+EXAMPLE_PIPELINE_DUCKDB = "example_pipeline.duckdb"
```
Comment (rexledesma) [inline python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt_tests/dlt_tests/test_op.py:8]: Probably want to make this a fixture and ensure that its memory is cleared before test invocation.

**Review by rexledesma — CHANGES_REQUESTED** [review]
See comments, but looks good overall

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt_tests/dlt_tests/test_op.py (hunk 3 above)
Comment (cmpadden) [reply python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt_tests/dlt_tests/test_op.py:8]: Created a separate PR with a refactor to used a shared fixture w/ setup / teardown.

https://github.com/dagster-io/dagster/pull/21199/files

Not entirely sure I like how I'm returning a tuple and have to destructure it, open to suggestions.

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py (hunk 1 above)
Comment (cmpadden) [reply python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py:137]: Idk why I struggle so much with the `check` module and type checking, but using `check.invariant` I get warnings downstream that the variable can possibly be none when using:

```
check.invariant(not (dlt_source and dlt_pipeline))
```

Am I making a silly boolean logic mistake? 😅 

```
>>> for (a, b) in [(None, None), (None, 1), (1, None), (1, 1)]:
...     try:
...         _ = check.invariant(not (a and b))
...         _c = True
...     except:
...         _c = False
...     print(a is None or b is None, _c)
...
True True
True True
True True
False False
```

Comment (cmpadden) [reply python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py]: I updated it to the following which works as expected.
```python
dlt_source = check.not_none(
    dlt_source, "dlt_source is a required parameter in an op context"
)
dlt_pipeline = check.not_none(
    dlt_pipeline, "dlt_pipeline is a required parameter in an op context"
)
```

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py [hunk 4]
```diff
@@ -99,26 +99,47 @@ def extract_resource_metadata(
...
+            dagster_dlt_translator = check.inst(
+                first_asset_metadata.get(META_KEY_TRANSLATOR), DagsterDltTranslator
+            )
 
-        dlt_source = check.inst(first_asset_metadata.get(META_KEY_SOURCE), DltSource)
-        dlt_pipeline = check.inst(first_asset_metadata.get(META_KEY_PIPELINE), Pipeline)
-        dagster_dlt_translator = check.inst(
-            first_asset_metadata.get(META_KEY_TRANSLATOR), DagsterDltTranslator
-        )
+        else:
```
Comment (rexledesma) [inline python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py:133]: nit: the `else` isn't needed, so just remove it.

**Review by rexledesma — APPROVED** [review]

**Pull request merged**